Add/caption for gallery block

tellthemachines opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

Added a caption for the gallery block. This caption refers to the gallery as a whole, so semantically it works by wrapping the whole gallery block in a `<figure>` element and adding a `<figcaption>` with the caption inside.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Tested locally, checked if previously created galleries still work, ran unit tests, updated fixtures.

Note: The caption on the editor side will likely have the same accessibility issues as the image caption. These are discussed in detail in [@talldan's table caption PR](https://github.com/WordPress/gutenberg/pull/15554#issuecomment-510725811)
I don't believe we have agreed on the best way forward for that, so it might be best to address it as a separate PR. 

## Screenshots <!-- if applicable -->

<img width="528" alt="Editor view of gallery with caption" src="https://user-images.githubusercontent.com/8096000/62855555-e8e8d480-bd35-11e9-9967-564dc6b03009.png">

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

mchowning committed:
[RNMobile] Reset toolbar scroll on content change

tellthemachines committed:
Basic gallery caption.

tellthemachines committed:
Basic gallery caption.

tellthemachines committed:
add block deprecation

tellthemachines committed:
add block deprecation

tellthemachines committed:
Regenerate test fixtures and fix failures

tellthemachines committed:
Regenerate test fixtures and fix failures

tellthemachines committed:
e2e snapshot update

tellthemachines committed:
e2e snapshot update

tellthemachines committed:
Fix editor styles

tellthemachines committed:
Fix editor styles

tellthemachines committed:
Fix block transforms e2e

tellthemachines committed:
Fix block transforms e2e

tellthemachines committed:
Workaround for failing tests

tellthemachines committed:
Workaround for failing tests

gziolo commented on package-lock.json:
```diff
@@ -7708,7 +7708,7 @@
 								"minimist": {
 									"version": "1.2.0",
-									"resolved": false,
+									"resolved": "",
```
This is surprising to see those changes.

tellthemachines replied:
Oh bugger, shouldn't have committed this file. I keep getting this diff when I run `npm install` even though I'm updating my npm version regularly 😩

tellthemachines replied:
So reverting this file to its previous state before my commit doesn't work because it's been updated in master in the meantime. I tried `nvm install --latest-npm ` followed by `npm install` and even `npm run docs:build` on my local branch and there's no diff. Is it possible the current `package-lock` on master wasn't generated with the latest npm?

gziolo replied:
Yes, if there is new npm version, and they changed something then we will have to patch `master`.

gziolo commented on packages/block-library/src/gallery/block.json:
```diff
@@ -48,6 +48,11 @@
 		"columns": {
 			"type": "number"
 		},
+		"galleryCaption": {
```
Can it be `caption`? It seems repetitive to include gallery as a prefix.

tellthemachines replied:
The problem is that gallery images already have `caption`s so it's less confusing if this one has a name that clearly indicates what it is a caption of.

gziolo replied:
It depends on where you look at, I found it confusing when seeing it in the `attributes` section `block.json` file. Gallery itself can have only one caption. In the future, we might refactor this block and convert all images to nested blocks at the issue you pointed out would solve on its own. You can always alias the caption of an image if it has to be referenced in the same scope or extract the image logic to its own component.

gziolo commented on packages/block-library/src/gallery/save.js:
```diff
@@ -46,6 +46,22 @@ export default function save( { attributes } ) {
+		</>
+	);
+
+	if ( galleryCaption ) {
```
If there is no change in the markup for the case when we don't provide the caption, do we really need to build all deprecation story? It seems like it should out of the box if the caption is not provided. I don't see any changes in the deprecated version serialized in the fixtures folder (input is almost identical to the output when you ignore whitespaces).

tellthemachines replied:
I'm not sure! The markup of the new version is the same but there's an added attribute. Happy to remove the deprecation if it's not needed though 🙂

gziolo replied:
You would have to tests it but I know we had similar cases in the past and we didn't have to build deprecation logic when new optional HTML attributes were added or wrapped with optional HTML tags. It seems like this is the case here.

gziolo commented on packages/block-library/src/gallery/style.scss:
```diff
@@ -1,8 +1,10 @@
-.wp-block-gallery {
+.wp-block-gallery,
```
In the case where Gallery block has a caption then it produces HTML where `.blocks-gallery-grid` is enclosed in `.wp-block-gallery`, won't it cause any issues?

tellthemachines replied:
I've tested this and added an override on the editor side for `figure.wp-block-gallery` to `display:block` as the flex display was causing issues. It seems to be fine on the post rendering side so far, but I can always add that override globally if needed. The other styles are pretty harmless.

gziolo commented on packages/block-library/src/gallery/style.scss:
```diff
@@ -147,4 +149,11 @@
 		}
 	}
+
+	+ figcaption {
```
Where those styles should get applied? To the gallery's caption only? If yes, then it should be placed outside of `.blocks-gallery-grid`.

tellthemachines replied:
Yup these are for the gallery caption only, and the problem here is that the RichText component won't allow me to add a classname to that element. Because there can potentially be other `figcaption`s inside the gallery, I'm targeting this one as `.blocks-gallery-grid + figcaption`. The plus sign between the two selectors makes it target only the figcaption that is the immediate next sibling to `.blocks-gallery-grid`.

tellthemachines committed:
Undo package-lock changes

tellthemachines committed:
Undo package-lock changes

tellthemachines committed:
Revert "Undo package-lock changes"

This reverts commit f4f8ba975a15f40fab3cfefded1ff4becc46e3cc.

talldan committed:
Update github action exit codes (#17002)

* Update github action exit codes

* Remove Filter Opened from github actions

tellthemachines committed:
Fix messed up styles.

tellthemachines committed:
Fix messed up styles.

tellthemachines committed:
Try to fix package-lock issue _again_

jasmussen commented:
Thanks for this! The caption field will be a delightful addition to the gallery.

For whatever reason I cannot see the caption on my local setup when I compile this, but this is almost certainly an issue on my end. Not sure what it is — I'm on the right branch, and I see the code changes in the gallery block, but when I select the gallery block I don't see the caption field popping out as it does on Image.

However a few things:

- This should mimic the caption of other blocks, such as Image and Video. That means `style.scss` should have only minimal stylings necessary for the block to function, i.e. layout.
- Additional visual styles, like the centering, the font size, should exist in a new file, `theme.scss`, which should contain something along the lines of:

```
.wp-block-gallery figcaption {
	@include caption-style-theme();
}
```

☝️ that up there is a mixin that includes the necessary CSS to style the caption the same as other block captions.

Make sense?

jorgefilipecosta committed:
chore: Update: Use mixins in button instead of media queries (#17012)

mapk commented:
I just tested this and it's coming along nicely! Thanks for the work here.

A couple design adjustments are needed:

1. As @jasmussen pointed out the styling for the caption placeholder and text should mimic the styling on the Image block. Notice the difference in font size. 

**Image caption**

![image-caption](https://user-images.githubusercontent.com/617986/62962066-f1cbbb80-bdb2-11e9-8b8a-2c45c918ec5b.png)

**Gallery caption**

![gallery-caption](https://user-images.githubusercontent.com/617986/62962086-fd1ee700-bdb2-11e9-8e65-36bcacb717de.png)

----

2. There's a large gap now between the Appender button and the images. And also a larger space on the right side of the images within the block.

**Master branch Gallery block**

![master](https://user-images.githubusercontent.com/617986/62962140-22abf080-bdb3-11e9-9a86-ebb8e767904f.png)

**This branch Gallery block**

![Screen Shot 2019-08-13 at 10 08 48 AM](https://user-images.githubusercontent.com/617986/62962302-80d8d380-bdb3-11e9-8735-53f4fc5876ea.png)

jorgefilipecosta committed:
Add editor-styles-wrapper class to the widgets screen block list (#17023)

jorgefilipecosta committed:
Add: Button Block appender to the widgets screen. (#16971)

gziolo committed:
Scripts: Add section about updating package after new releases (#17026)

* Scripts: Add section about updating package after new releases
It also contains some editorial changes to make things clearer

* Use one version of apostrophe

tellthemachines committed:
Addressed design feedback.

tellthemachines committed:
Addressed design feedback.

jorgefilipecosta committed:
Fix: Duplicate button appears even if the block is not allowed (#17007)

tellthemachines committed:
Change caption name and remove deprecation.

tellthemachines committed:
Change caption name and remove deprecation.

tellthemachines commented:
Thanks for the feedback @jasmussen and @mapk, I have now addressed all the styling issues. However, themes don't yet include styles for gallery captions yet so there is still a visual discrepancy with the image caption. Not sure what the best procedure is in this situation.

jasmussen commented:
Thanks for addressing these things! I have to jump onto a different project so don't have time to compile the branch, but the code looks good. And to answer your question:

> However, themes don't yet include styles for gallery captions so there is still a visual discrepancy with the image caption. Not sure what the best procedure is in this situation.

I agree this isn't super ideal, but having discussed this in the past in similar contexts around the Image and Embed blocks, this is on the theme developer to either address themselves (through styling the `figcaption` globally), or to opt into "opinionated block styles", which causes the theme.scss file to load.

karmatosed commented:
> However, themes don't yet include styles for gallery captions so there is still a visual discrepancy with the image caption. Not sure what the best procedure is in this situation.

Like @jasmussen I think it's ok for themes to support from now on. We could do bring it more inline with the image caption, that would be my feelings that it should look as close as possible and inherit any styling for that.

youknowriad committed:
Fix the double scrollbar appearing in full screen mode (#17031)

epiqueras committed:
Core Data: Return updated record in `saveEntityRecord`. (#17030)

youknowriad committed:
Fix performance tests with the introduction of the navigation mode (#17034)

ellatrix committed:
RichText: ignore selection changes during composition (#16960)

youknowriad committed:
Bump plugin version to 6.3.0

ellatrix committed:
RichText 🧹 (clean up, move Autocomplete, remove DOM dependency) (#16905)

* RichText 🧹

* Clean up

* Clean up

* Simplify

* Move Autocomplete

* children => Editable wrapper component

* Bind Editable wrapper to class

* Minimise diff

ellatrix committed:
RichText (block context): move list controls to list block (#16962)

* Move list controls

* Don't restrict to isSelected

tellthemachines commented:
> We could do bring it more inline with the image caption, that would be my feelings that it should look as close as possible and inherit any styling for that.

Right now it has the same styles as the image caption, but on certain themes, such as Twenty Nineteen, the theme styles targeting the image caption are too specific to apply to the gallery caption. For some reason on Twenty Nineteen the font size for gallery caption is tiny. But on other themes that apply more generic styles it looks OK.

jasmussen commented:
> Right now it has the same styles as the image caption, but on certain themes, such as Twenty Nineteen, the theme styles targeting the image caption are too specific to apply to the gallery caption. For some reason on Twenty Nineteen the font size for gallery caption is tiny. But on other themes that apply more generic styles it looks OK.

@kjellr any thoughts here? Is TwentyNineteen styling the `figcaption` directly?

nosolosw committed:
Update notifications (#17046)

Jackie6 committed:
List block: add start and reversed attributes (#15113)

* Squash for easier merge:
f6ad4f07e6a1c8d99e55aa7acb01cf8ef1dfddd7 (HEAD -> update/list-block, Jackie6/update/list-block) Add id to BaseControl
212cab106044c4d7300d9441e48241b911db0bcb Fix unit test and e2e test
421a198bf58431c91b4c09d6ee45ba932762cb15 Refine the type definition of start and reversed
4172707059558ba9fc47d9557457e0b8e441f050 Fix typo
21ed1dbeb7881a9a5a6eea85fb3293e9d6ba0d5a Update CSS styles
925f1ab1ed073b31358a129c62abd56021e14c38 Update list type style and remove description
ef0aca95cb88203ceff78b81473e37ad8fc22714 Allow the start to be NaN
495166d3fc14591ceb88597915a19e976b2ebbeb Change to check strict equality
216ca51cfacfc174ca00d9669b88b53eaa6d0356 Fix the accidental change of travis yml
8e31259624ece930c385bf9cb7a28046505c7aff Change local state to attributes
09199ff9516aab099e47ed6e61fd4e2fadfdb95b Add list type, start, reversed settings

* Revert adding  attribute for now, let's extract to a separate PR

* Use proper TextControl

* Adjust toggle control to unset attribute

* Clearer label

getdave committed:
Adds default funcs as props (#17036)

Previously the component assumed that the onChange and onInput props would always be present. However, there are times where this may not be the case such as when the BlockPreview component uses the provider to render the previews. In such cases `onChange` is called but is undefined which causes an error to be thrown.

To avoid this we provide lodash noop as defaults for both “change” props.

ellatrix commented on packages/block-library/src/gallery/save.js:
```diff
@@ -46,6 +46,22 @@ export default function save( { attributes } ) {
+	);
+
+	if ( galleryCaption ) {
+		return (
```
Could we always wrap in a `figure` tag, for consistency? We do the same for all other media (image, audio, video, embed...)

ellatrix commented:
Love it, but I think we should always be wrapping in a `figure` tag, whether there is a caption or not. See other media blocks. :)

bfintal committed:
Added ESNext examples to format API tutorial (page 1) (#16804)

* Added ESNext examples

* Added missing end tag

bfintal committed:
Added ESNext examples to format API tutorial (page 2) (#16803)

* Added ESNext examples

* Added missing end tags

bfintal committed:
Added ESNext examples to format API tutorial (page 3) (#16802)

* Added ESNext examples

* Added missing end tag

* Use wp.blockEditor instead of wp.editor

mapk commented:
I agree that some theme are going to have some difficulties here. They'll need to add styles for it. As for how it works right now, the appearance in the editor with the Gutenberg Starter Theme is exactly how I imagined it should look. 

![Screen Shot 2019-08-15 at 6 20 15 PM](https://user-images.githubusercontent.com/617986/63137061-8ab42f80-bf89-11e9-8a80-0401085c6cec.png)

This looks great from a design perspective. 👍

talldan commented:
Would be great to add a simple e2e test here for the caption:
https://github.com/WordPress/gutenberg/tree/master/packages/e2e-tests/specs/blocks

gziolo committed:
Scripts: Improve the way test files are discovered (#17033)

* Scripts: Improve the way test files are discovered

* Fix file matching patterns and file extensions for integration tests

jorgefilipecosta committed:
Fix: Button block does not centers on the editor (#17063)

mchowning committed:
[RNMobile] Extract caption component (#16825)

* Extract caption component

* Use caption component for video

* Move Caption to block-editor for mobile tests

This move avoids a dependency issue that was causing the mobile tests to
fail due to the Caption component's import of RichText from the
block-editor package.

* Pass onBlur to Caption component

* Connect Caption component to store

donmhico committed:
Update: Latest Posts Block: Use (no title) instead of (Untitled) for untitled posts (#17074)

tellthemachines committed:
Revert "Change caption name and remove deprecation."

This reverts commit 9a0876c608eb3eaaeab51189ee1d34f665cf336d.

tellthemachines committed:
Revert "Change caption name and remove deprecation."

This reverts commit 9a0876c608eb3eaaeab51189ee1d34f665cf336d.

tellthemachines committed:
Re-add gallery caption name change.

tellthemachines committed:
Re-add gallery caption name change.

tellthemachines committed:
Merge branch 'add/caption-for-gallery-block' of github.com:WordPress/gutenberg into add/caption-for-gallery-block

tellthemachines commented:
Closing this PR because I messed up the rebase, will have updated one up soon.
